mantidproject/mantid
# Refactor ReflMainView and ReflMainViewPresenter

This is the first step in #14732.
- `ReflMainView` has been split into two different classes: `ReflMainView` and `ReflTableView`. `ReflTableView` has the functionality related to the processing table: options to open/save a table, add/delete rows, etc, even the "Reflectometry UI Options" have been transferred to `ReflTableView` as they refer to the processing table.
- Each class has its own presenter, and each presenter has its own unit tests:
  - `ReflMainView` -> `ReflMainViewPresenter` -> `ReflMainViewPresenterTest`
  - `ReflTableView` -> `ReflTableViewPresenter` -> `ReflTableViewPresenterTest`
- Files defining the mock objects have also been split:
  - `ReflMainViewMockObjects`
  - `ReflTableViewMockObjects`

The following diagram tries to represent the new structure of the interface (apologies if it does not follow the standards, but I hope it is clear enough):

![newdiagram](https://cloud.githubusercontent.com/assets/8956985/13775106/a5cf2bc0-ea9a-11e5-83f1-758d170b535e.png)
- The left section represents reflectometry related stuff, which includes the abstract presenter `IReflPresenter` defining the notification interface, the concrete presenter `ReflMainViewPresenter`, which does all the UI logic, and interacts with the abstract view, `ReflMainView`. 
- The concrete view, `QtReflMainView` has a component `QReflTableView`. This is now a new widget independent from `QtReflMainView` (although some communication between them is needed).
- The right section reseprents all the things that should be unrelated to reflectometry (at the moment, they are still related to this specific area, but they will be generalized in the next issue). Essentially it consists of an abstract view, `ReflTableView`, that is updated by a `ReflTableViewPresenter` implementing an abstract `IReflTablePresenter`.

**To test:**
1. Please consider the above diagram and the following comments and make sure I am following the correct approach:
- Some changes are visible to users: most of the funcionality has been moved to `ReflTable` (by `ReflTable` I mean both the view and presenter). This means that most of the options in the menus have been removed and moved to the tool bar. Actually, only the `Slit Calculator` remains in the `Reflectometry` menu bar.
- The existing progress bar now belongs to `ReflTable`. I've added a new progress bar in the `Search Runs` section which will report progress when transferring runs. The original progress bar will only report processing. One thing to consider is that now we have two different `ReflProgress` classes defined in each presenter. See [here](https://github.com/mantidproject/mantid/blob/14921_Refactor_ReflMainView_and_ReflMainViewPresenter/MantidQt/CustomInterfaces/src/Reflectometry/ReflMainViewPresenter.cpp#L28-L52) and [here](https://github.com/mantidproject/mantid/blob/14921_Refactor_ReflMainView_and_ReflMainViewPresenter/MantidQt/CustomInterfaces/src/Reflectometry/ReflTableViewPresenter.cpp#L44-L68). Should we merge them into a single class and move it to a separate header that can be used by both presenters? If so, in which directory or subdirectory should this class be?
- Some communication between `QtReflMainView` and `QReflTableView` is needed, specifically in the following cases:
  - When transferring runs to the table, `QtReflMainView` needs to get from its presenter the set of runs to transfer and then send them to `QReflTableView` so they can be added to the table. This is done [here](https://github.com/mantidproject/mantid/blob/14921_Refactor_ReflMainView_and_ReflMainViewPresenter/MantidQt/CustomInterfaces/src/Reflectometry/QtReflMainView.cpp#L161-L164).
  - Instrument combo boxes need to be synchronized, whenever one changes, the other should be updated accordingly. This is done [here](https://github.com/mantidproject/mantid/blob/14921_Refactor_ReflMainView_and_ReflMainViewPresenter/MantidQt/CustomInterfaces/src/Reflectometry/QtReflMainView.cpp#L52-L58)
  - `QReflTableView` needs to run python code, for instance when the user loads a TBL file or plots a row. As it is not a `UserSubWindow` anymore, but a `MantidWidget`, when the signal `runAsPythonScript(const QString &, bool)` is emitted, it has to be re-emitted by `QtReflMainView`. This is done [here](https://github.com/mantidproject/mantid/blob/14921_Refactor_ReflMainView_and_ReflMainViewPresenter/MantidQt/CustomInterfaces/src/Reflectometry/QtReflMainView.cpp#L60-L62)
1. Code review
2. Please make sure all the existing functionality still works (transferring runs, processing, importing/exporting a TBL, plotting rows and groups). Most of it is tested in unit tests.

Fixes #14921 

Release notes have been updated.

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
